### PR TITLE
Fix openstack inventory.py looping

### PR DESCRIPTION
**Pull request by michaelgugino**
Currently, performance of this inventory script is quite
slow. This is due to looping over the same large list of
hosts repeatedly.

This commit refactors the inventory script to only loop
over the list of hosts once.  This will provide great
performance improvements for large clusters.

Replaces: https://github.com/openshift/openshift-ansible/pull/7581

### CODE REVIEW

**Comment by mbruzek** [comment]
Hello @michaelgugino I too was concerned about the multiple loops and created a new inventory.py file to reduce the number of loops over the same server listing [0]. The code I came up with looks very similar to your change.

However, when I tested the performance of my change vs the original inventory.py on a running cluster, I found that it made no measurable difference at all. My test cases were 100 nodes, and 700 nodes and I used time to get the real time it took to run the command. My measurements fluctuated around the same real time as the original inventory.py (and I have the spreadsheets to document this). 

My theory is my change made no time improvement because the in-memory looping is very fast when compared to the API call to return all the servers from shade.

[0] - http://pastebin.test.redhat.com/568339

File: playbooks/openstack/inventory.py [hunk 1]
```diff
@@ -16,52 +16,72 @@
...
+            group = server.metadata.get('group')
+            if group not in inventory:
+                inventory[group] = {'hosts': []}
+            inventory[group]['hosts'].append(server.name)
 
-    etcd = [server.name for server in cluster_hosts
-            if server.metadata['host-type'] == 'etcd']
-    if not etcd:
-        etcd = masters
+        inventory['cluster_hosts'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: I think this should be `inventory['cluster_hosts']['hosts'].append(server.name)`

File: playbooks/openstack/inventory.py [hunk 2]
```diff
@@ -16,52 +16,72 @@
...
 
-    infra_hosts = [server.name for server in cluster_hosts
-                   if server.metadata['host-type'] == 'node' and
-                   server.metadata['sub-host-type'] == 'infra']
+        if server.metadata['host-type'] == 'master':
+            inventory['masters'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: This is missing `['hosts']` as well

File: playbooks/openstack/inventory.py [hunk 3]
```diff
@@ -16,52 +16,72 @@
...
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    app = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'node' and
-           server.metadata['sub-host-type'] == 'app']
+        if server.metadata['host-type'] == 'etcd':
+            inventory['etcd'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: And here

File: playbooks/openstack/inventory.py [hunk 4]
```diff
@@ -16,52 +16,72 @@
...
+            osev3.add(server.name)
+        if not inventory['etcd']:
+            inventory['etcd'] = inventory['masters']
 
-    cns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'cns']
+        if (server.metadata['host-type'] == 'node' and
+                server.metadata['sub-host-type'] == 'infra'):
+            inventory['infra_hosts'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: Ditto

File: playbooks/openstack/inventory.py [hunk 5]
```diff
@@ -16,52 +16,72 @@
...
+            nodes.add(server.name)
 
-    nodes = list(set(masters + infra_hosts + app + cns))
+        if (server.metadata['host-type'] == 'node' and
+                server.metadata['sub-host-type'] == 'app'):
+            inventory['app'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: Ditto

File: playbooks/openstack/inventory.py [hunk 6]
```diff
@@ -16,52 +16,72 @@
...
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    dns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'dns']
+        if server.metadata['host-type'] == 'cns':
+            inventory['cns'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: Ditto

File: playbooks/openstack/inventory.py [hunk 7]
```diff
@@ -16,52 +16,72 @@
...
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    load_balancers = [server.name for server in cluster_hosts
-                      if server.metadata['host-type'] == 'lb']
+        if server.metadata['host-type'] == 'dns':
+            inventory['dns'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: Ditto

File: playbooks/openstack/inventory.py [hunk 8]
```diff
@@ -16,52 +16,72 @@
...
 
-    osev3 = list(set(nodes + etcd + load_balancers))
+        if server.metadata['host-type'] == 'lb':
+            inventory['load_balancers'].append(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: Ditto

**Comment by mbruzek** [comment]
> Best way would to be to filter initial call to cloud.list_servers
If we could filter on master, cns, and other categories that would be ideal! I looked into filtering the list_servers() call as well, but **could not** find a way to filter on metadata the key we are checking for group membership. From the OpenStack docs: "There is whitelist for valid filter keys. Any filter key other than from whitelist will be silently ignored." 

I found some documentation that did not list "metadata" as a filterable object as admin or as regular user [0]. Search for "filter".

It would be better if we could ask shade for all the groups of servers that we need to build the inventory response. Could we base the group on a "tag" or something else we could filter on? If you have any ideas or other thoughts on how to query shade for these groups this please let me know.

[0] - https://developer.openstack.org/api-ref/compute/#list-servers

File: playbooks/openstack/inventory.py [hunk 9]
```diff
@@ -16,52 +16,72 @@
...
+            osev3.add(server.name)
+        if not inventory['etcd']['hosts']:
+            inventory['etcd']['hosts'] = inventory['masters']['hosts']
 
-    cns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'cns']
+        if (server.metadata['host-type'] == 'node' and
+                server.metadata['sub-host-type'] == 'infra'):
+            inventory['infra_hosts']['hosts'].append(server.name)
+            nodes.add(server.name)
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: I think this also needs `osev3.add(server.name)`

File: playbooks/openstack/inventory.py [hunk 10]
```diff
@@ -16,52 +16,72 @@
...
+            osev3.add(server.name)
 
-    app = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'node' and
-           server.metadata['sub-host-type'] == 'app']
+        if server.metadata['host-type'] == 'etcd':
+            inventory['etcd']['hosts'].append(server.name)
+            osev3.add(server.name)
+        if not inventory['etcd']['hosts']:
+            inventory['etcd']['hosts'] = inventory['masters']['hosts']
```
Comment (tzumainn) [inline playbooks/openstack/inventory.py]: I think this check belongs outside the for loop.

**Comment by michaelgugino** [comment]
@mbruzek 
Yeah, looks like we can only filter with tags, not meta.

We could override this method: https://github.com/openstack-infra/shade/blob/master/shade/openstackcloud.py#L2183

and make it multiprocess with map_async.  That would probably speed up the results significantly because it looks to be applied serially.

**Comment by tzumainn** [comment]
Thanks for making the fixes! @mbruzek, I know that your testing showed that this sort of change didn't make a ton of difference; but it looks like it makes some, and the additional changes you guys are discussion seem independent of this. Are you okay with this merging?

File: playbooks/openstack/inventory.py [hunk 11]
```diff
@@ -130,24 +152,14 @@ def build_inventory():
         server for server in cloud.list_servers()
```
Comment (mbruzek) [inline playbooks/openstack/inventory.py]: Could the `cluster_hosts` loop be included in the `base_openshift_inventory()` function as well? Pass in the server list and return an inventory and/or the cluster_hosts list (if necessary)?

File: playbooks/openstack/inventory.py [hunk 12]
```diff
@@ -16,52 +16,74 @@
...
 
-    dns = [server.name for server in cluster_hosts
-           if server.metadata['host-type'] == 'dns']
+        if server.metadata['host-type'] == 'dns':
+            inventory['dns']['hosts'].append(server.name)
 
-    load_balancers = [server.name for server in cluster_hosts
-                      if server.metadata['host-type'] == 'lb']
+        if server.metadata['host-type'] == 'lb':
+            inventory['load_balancers']['hosts'].append(server.name)
```
Comment (mbruzek) [inline playbooks/openstack/inventory.py]: The inventory group key should be "lb" here to remain consistent with the previous code.

File: playbooks/openstack/inventory.py [hunk 13]
```diff
@@ -16,52 +16,74 @@
...
+            inventory['masters']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    infra_hosts = [server.name for server in cluster_hosts
-                   if server.metadata['host-type'] == 'node' and
-                   server.metadata['sub-host-type'] == 'infra']
+        if server.metadata['host-type'] == 'etcd':
+            inventory['etcd']['hosts'].append(server.name)
+            osev3.add(server.name)
```
Comment (mbruzek) [inline playbooks/openstack/inventory.py]: Does this mean that etcd servers are **not** part of the nodes set? It looks like their hosts will be added to inventory a few lines down, but will they/should they be in the node set as well?

**Comment by michaelgugino** [comment]
> Not opposed to this kind of change at all. But I did want to point out it can not be a replacement for #7581 because it will be returning all the nodes all the time which is expensive and time consuming for our large scale up operations. These are separate issues and addressed with separate PRs.

I'm not sure #7581 will have the desired effect as-is.  Unfortunately, I don't have an OS cluster handy to test.

**Pull request closed without merging**